### PR TITLE
build: update to TypeScript 3.4

### PR DESCRIPTION
**Pull request by filipesilva**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [x] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the new behavior?
Update to TS 3.4: https://devblogs.microsoft.com/typescript/announcing-typescript-3-4-rc/

## Does this PR introduce a breaking change?

- [x] Yes
- [ ] No

### CODE REVIEW

**Comment by filipesilva** [comment]
Similar to https://github.com/angular/angular/pull/29004

**Comment by filipesilva** [comment]
Related to https://github.com/angular/tsickle/pull/1004

File: packages/compiler-cli/src/ngcc/src/rendering/renderer.ts [hunk 1]
```diff
@@ -91,7 +91,7 @@ export abstract class Renderer {
     const renderedFiles: FileInfo[] = [];
 
     // Transform the source files.
-    this.bundle.src.program.getSourceFiles().map(sourceFile => {
+    this.bundle.src.program.getSourceFiles().forEach(sourceFile => {
```
Comment (rkirov) [inline packages/compiler-cli/src/ngcc/src/rendering/renderer.ts]: what happened here?

Comment (filipesilva) [reply packages/compiler-cli/src/ngcc/src/rendering/renderer.ts]: TS was showing an error here that the return value was not used: https://circleci.com/gh/angular/angular/248206

```
packages/compiler-cli/src/ngcc/src/rendering/renderer.ts:94:5 - error TS21222: return value is unused.
	See http://tsetse.info/check-return-value

 94     this.bundle.src.program.getSourceFiles().map(sourceFile => {
        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 95       const compiledFile = decorationAnalyses.get(sourceFile);
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
... 
101       }
    ~~~~~~~
102     });
```

Comment (rkirov) [reply packages/compiler-cli/src/ngcc/src/rendering/renderer.ts]: Interesting that TS3.4 will trigger that, it was not used before too.

File: aio/src/app/shared/reporting-error-handler.ts [hunk 2]
```diff
@@ -28,10 +28,12 @@ export class ReportingErrorHandler extends ErrorHandler {
   }
 
   private reportError(error: string | Error) {
-    if (typeof error === 'string') {
-      this.window.onerror(error);
-    } else {
-      this.window.onerror(error.message, undefined, undefined, undefined, error);
+    if (this.window.onerror) {
```
Comment (filipesilva) [inline aio/src/app/shared/reporting-error-handler.ts]: TS 3.4 was failing compilation with the error below https://circleci.com/gh/angular/angular/248678

```
ERROR in src/app/shared/reporting-error-handler.ts(32,7): error TS2721: Cannot invoke an object which is possibly 'null'.
src/app/shared/reporting-error-handler.ts(34,7): error TS2721: Cannot invoke an object which is possibly 'null'.
```

File: aio/tests/e2e/onerror.e2e-spec.ts [hunk 3]
```diff
@@ -176,21 +203,24 @@ createDirectiveInstance@core.umd.j`);
...
-  });
+     });
 
   async function callOnError(message, url, line, column, error) {
     await browser.executeScript(function() {
       // reset the ga queue
       (window as any).ga.q.length = 0;
       // post the error to the handler
-      window.onerror(arguments[0], arguments[1], arguments[2], arguments[3], arguments[4]);
+      if (window.onerror) {
```
Comment (filipesilva) [inline aio/tests/e2e/onerror.e2e-spec.ts]: TS was failing compilation with the error below https://circleci.com/gh/angular/angular/248820

```
[18:29:57] E/launcher - Error: TSError: ⨯ Unable to compile TypeScript
tests/e2e/onerror.e2e-spec.ts (193,7): Cannot invoke an object which is possibly 'null'. (2721)
```

File: aio/tests/e2e/onerror.e2e-spec.ts [hunk 4]
```diff
@@ -1,5 +1,5 @@
-import { browser } from 'protractor';
-import { SitePage } from './app.po';
+import {browser} from 'protractor';
```
Comment (filipesilva) [inline aio/tests/e2e/onerror.e2e-spec.ts]: Saving this file triggered the clang auto format. I could also save without formatting, but that begs the question of why this wasn't formatted to begin with.

File: packages/compiler-cli/src/ngcc/src/rendering/renderer.ts (hunk 1 above)
Comment (filipesilva) [reply packages/compiler-cli/src/ngcc/src/rendering/renderer.ts]: Interesting indeed. Now that I look at it with fresh eyes, this isn't a TypeScript error proper, but a TseTse error: https://github.com/bazelbuild/rules_typescript/blob/b4bc468bfb03c05b68dde0b4f846bcd1a8fb7b01/internal/tsetse/tests/check_return_value/BUILD#L19

I wonder why TseTse wasn't catching this before. 

cc @alexeagle

File: aio/tests/e2e/onerror.e2e-spec.ts (hunk 4 above)
Comment (gkalpak) [reply aio/tests/e2e/onerror.e2e-spec.ts]: The files under `aio/` are not formatted with `clang-format` (when using `gulp format` for example).
Some of the files definitely should not be formatted (e.g. `aio/content/examples/`, which should follow the Angular Style Guide) and although the app files themselves (in `aio/src/`) could be formatted, they historically haven't been :grin:

(Note: The annoying auto-formatting will go away, once I find some time to clean-up #28842 😇)

Comment (filipesilva) [reply aio/tests/e2e/onerror.e2e-spec.ts]: Saved without formatting then, thanks for the context.

**Comment by RinMinase** [comment]
[Typescript 3.4.1](https://github.com/Microsoft/TypeScript/releases) is out, but your [package dependencies file](https://github.com/filipesilva/angular/blob/b9bb9df699aa8544864bd2f161f5dd12fbf79756/package.json#L101) states `~3.4.0-rc`, would this be applicable to use `^3.4.1` or would this introduce another set of breaking changes?

File: packages/compiler-cli/src/ngcc/src/rendering/renderer.ts (hunk 1 above)
Comment (alexeagle) [reply packages/compiler-cli/src/ngcc/src/rendering/renderer.ts]: my guess is that we only look for usages of `Array#map`, while `program.getSourceFiles()` maybe was some other iterable type before like `ReadonlyArray` I guess now it became assignable to `Array`?

**Comment by filipesilva** [comment]
@RinMinase this PR was just an initial test when the RC came out to see how much work would be necessary on this repository. We will change it to support a stable version before merging.

File: aio/package.json [hunk 5]
```diff
@@ -157,7 +157,7 @@
     "tree-kill": "^1.1.0",
     "ts-node": "^3.3.0",
     "tslint": "~5.9.1",
-    "typescript": "~3.3.3333",
+    "typescript": "~3.4.0-rc",
```
Comment (alexeagle) [inline aio/package.json]: we are ready to merge this, could you update this to the released 3.4?

**Comment by alexeagle** [comment]
caretaker Igor: on your plate to clean this up and check the peerDep ranges everywhere

File: integration/dynamic-compiler/package.json [hunk 6]
```diff
@@ -5,7 +5,8 @@
   "scripts": {
     "build": "npm run clean && npm run ngc && npm run rollup && npm run rollup:lazy && npm run es5 && npm run es5:lazy",
     "clean": "rm -rf dist",
-    "es5": "tsc --target es5 --skipLibCheck --allowJs dist/bundle.es2015.js --out dist/bundle.js",
+    "//es5-note": "`--typeRoots ./dummy_folder_name/@types` is a workaround for https://github.com/Microsoft/TypeScript/issues/30845",
+    "es5": "tsc --target es5 --skipLibCheck --allowJs dist/bundle.es2015.js --out dist/bundle.js --typeRoots ./dummy_folder_name/@types",
```
Comment (filipesilva) [inline integration/dynamic-compiler/package.json]: `--typeRoots ./dummy_folder_name/@types` is a workaround for https://github.com/Microsoft/TypeScript/issues/30845.

This test does not depend on `@types/node`, but they are present in the root node_modules, both via a direct dependency and transitively through the `rollup` dependency.

**Review by alexeagle — APPROVED** [review]
thanks @filipesilva nice work 🌹

File: integration/bazel-schematics/test.sh [hunk 7]
```diff
@@ -30,7 +30,7 @@ function testBazel() {
   yarn
   # Force more recent TS version until new Angular CLI projects also use it.
   # --ignore-scripts is necessary because there is a postinstall script that uses ngc.
-  yarn add typescript@3.3.3333 --dev --ignore-scripts
+  yarn add typescript@3.4.0-rc --dev --ignore-scripts
```
Comment (IgorMinar) [inline integration/bazel-schematics/test.sh]: Why not 3.4.2? Did you miss to update this file?

**Review by IgorMinar — APPROVED** [review]
Other than the one missed version everything else looks great. Thank you.

File: integration/bazel-schematics/test.sh (hunk 7 above)
Comment (filipesilva) [reply integration/bazel-schematics/test.sh]: Yes, I missed this one, thank you for catching it 👍 

Fixed now.

**Pull request closed without merging**